File: uq/random/lhs.py
"""Latin Hypercube Sampling.

Author:
    Ilias Bilionis

Date:
    12/2/2012

"""

import numpy as np
import uq.core


def lhs_seed():
    """Return a random seed that can be used in lhs."""
    return uq.core.get_seed()

def lhs(n, k, seed=lhs_seed()):
    """Fill an n x k array with latin hyper-cube samples and return it.

    Arguments:
        n       ---     Number of samples.
        k       ---     Number of dimensions.

    Keyword Arguments:
        seed    ---     The random seed.

    """
    X = np.ndarray((k, n), order='F')
    uq.core.lhs(X, seed)
    return X.T

File: src/_lhs.cpp
// Latin hypercube sampling
// Author: Ilias Bilionis
// Date: 12/2/2012

#include <latin_center_dataset.hpp>
#include <boost/python.hpp>
#include <numpyconfig.h>
#include <arrayobject.h>
#include <numpy_interface.hpp>

using namespace boost::python;

inline void lhs(numeric::array& X, int seed)
{
    numpy_array<double> nX(X);
    latin_center(nX.shape[0], nX.shape[1], &seed, nX.data);
}

BOOST_PYTHON_MODULE(_lhs)
{
    import_array();
    numeric::array::set_module_and_type("numpy", "ndarray");

    def("lhs", lhs);
    def("get_seed", get_seed);
}
